tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -40,169 +41,219 @@ _decompress_chunk_init(void)
 
-	path->decompression_map = NIL;
+	/*
+	 * Fill the helper array of compressed attno -> compression info.
+	 */
+	FormData_hypertable_compression **compressed_attno_to_compression_info =
+		palloc0(sizeof(void *) * (path->info->compressed_rel->max_attr + 1));
+	ListCell *lc;
+	foreach (lc, path->info->hypertable_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		AttrNumber compressed_attno =
+			get_attnum(path->info->compressed_rte->relid, NameStr(fd->attname));
 
-	/* add count column */
-	tle = make_compressed_scan_meta_targetentry(path,
-												COMPRESSION_COLUMN_METADATA_COUNT_NAME,
-												DECOMPRESS_CHUNK_COUNT_ID,
-												list_length(scan_tlist) + 1);
-	scan_tlist = lappend(scan_tlist, tle);
+		if (compressed_attno == InvalidAttrNumber)
+		{
+			elog(ERROR,
+				 "column '%s' not found in the compressed chunk '%s'",
+				 NameStr(fd->attname),
+				 get_rel_name(path->info->compressed_rte->relid));
+		}
 
-	/* add sequence num column */
-	if (path->needs_sequence_num)
-	{
-		tle = make_compressed_scan_meta_targetentry(path,
-													COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME,
-													DECOMPRESS_CHUNK_SEQUENCE_NUM_ID,
-													list_length(scan_tlist) + 1);
-		scan_tlist = lappend(scan_tlist, tle);
+		compressed_attno_to_compression_info[compressed_attno] = fd;
 	}
 
-	/* check for system columns */
-	bit = bms_next_member(attrs_used, -1);
-	if (bit > 0 && bit + FirstLowInvalidHeapAttributeNumber < 0)
+	/*
+	 * Go over the scan targetlist and determine to which output column each
+	 * scan column goes.
+	 */
+	path->decompression_map = NIL;
+	foreach (lc, scan_tlist)
 	{
-		/* we support tableoid so skip that */
-		if (bit == TableOidAttributeNumber - FirstLowInvalidHeapAttributeNumber)
-			bit = bms_next_member(attrs_used, bit);
+		TargetEntry *target = (TargetEntry *) lfirst(lc);
+		if (!IsA(target->expr, Var))
+		{
+			elog(ERROR, "compressed scan targetlist entries must be Vars");
+		}
 
-		if (bit > 0 && bit + FirstLowInvalidHeapAttributeNumber < 0)
-			elog(ERROR, "transparent decompression only supports tableoid system column");
-	}
+		Var *var = (Var *) target->expr;
+		Assert(var->varno == path->info->compressed_rel->relid);
+		AttrNumber compressed_attno = var->varattno;
 
-	/* check for reference to whole row */
-	if (bms_is_member(0 - FirstLowInvalidHeapAttributeNumber, attrs_used))
-	{
-		ListCell *lc;
-		AttrNumber ht_attno = 0;
+		if (compressed_attno == InvalidAttrNumber)
+		{
+			/*
+			 * We shouldn't have whole-row vars in the compressed scan tlist,
+			 * they are going to be built by final projection of DecompressChunk
+			 * custom scan.
+			 * See compressed_rel_setup_reltarget().
+			 */
+			elog(ERROR, "compressed scan targetlist must not have whole-row vars");
+		}
 
-		foreach (lc, path->info->ht_rte->eref->colnames)
+		char *column_name = get_attname(path->info->compressed_rte->relid,
+										compressed_attno,
+										/* missing_ok = */ false);
```

Comment:
Nit: const-correctness

```suggestion
		const char *column_name = get_attname(path->info->compressed_rte->relid,
										compressed_attno,
										/* missing_ok = */ false);
```

Reply:
Fixed. By the way, the return value isn't const, it's an pstrdup'd string. If you mean something like 'static single assignment is generally good', I agree, but this also applies to another ten places in this PR.

Reply:
No, `const` data is not really constant, but it avoids mistakes like trying to modify values that are logically immutable in the function. It just serves as a signal to the compiler that the value is not intended to be modified. Might have avoided the issue we had with modifying partials in https://github.com/timescale/timescaledb/pull/4219, for example.